compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -43,46 +43,70 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
+        // Checks if all conditions are CALL EQEQ(_, _)
+        if (!areComparisons(calls))
             return null
 
-        // Subject should be the same for all conditions. Let's pick the first.
-        val subject = callToLabels[0].call.getValueArgument(0)!! as IrGetValue
-
-        // Don't generate repeated cases, which are unreachable but allowed in Kotlin.
-        // Only keep the first encountered case:
-        val cases =
-            callToLabels.map { ValueToLabel((it.call.getValueArgument(1) as IrConst<*>).value, it.label) }.distinctBy { it.value }
-
-        // Remove labels and "then expressions" that are not reachable.
-        val reachableLabels = HashSet(cases.map { it.label })
-        expressionToLabels.removeIf { it.label !in reachableLabels }
-
+        // Determine if all comparisons are between a single temporary variable
+        // and constants of a particular type: CALL EQEQ(tmp, constant)
         return when {
-            areConstIntComparisons(calls) ->
+            // UInts are an exceptional case: the IR contains CALL <unsafe-coerce><UInt, Int> around both arguments
+            // to the comparison. We check that structure in `areConstUIntComparison`.
+            areConstUIntComparisons(calls) -> {
+                val (subject, expressionToLabels, cases) = extractUIntSwitchComponents(callToLabels, expressionToLabels)
+                IntSwitch(
+                    subject,
+                    elseExpression,
+                    expressionToLabels,
+                    cases
+                )
+            }
+            // All remaining switchable types (Int and String) are primitives and can be handled uniformly.
+            !areTmpVariableVSConstantComparisons(calls) ->
+                return null
+            areConstIntComparisons(calls) -> {
+                val (subject, expressionToLabels, cases) = extractSwitchComponents(callToLabels, expressionToLabels)
                 IntSwitch(
                     subject,
                     elseExpression,
                     expressionToLabels,
                     cases
                 )
-            areConstStringComparisons(calls) ->
+            }
+            areConstStringComparisons(calls) -> {
+                val (subject, expressionToLabels, cases) = extractSwitchComponents(callToLabels, expressionToLabels)
                 StringSwitch(
                     subject,
                     elseExpression,
                     expressionToLabels,
                     cases
                 )
+            }
             else -> null // TODO: Enum, etc.
         }?.genOptimizedIfEnoughCases()
     }
 
-    private fun areConstComparisons(conditions: List<IrCall>): Boolean {
-        // All branches must be CALL 'EQEQ(Any?, Any?)': Boolean
-        if (conditions.any { it.symbol != codegen.classCodegen.context.irBuiltIns.eqeqSymbol })
-            return false
+    private fun areConstUIntComparisons(conditions: List<IrCall>): Boolean {
```

Comment:
Do we need to check each condition to see if they are unsafe-coerce calls? Could we just check if the LHS is an unsafe-coerce call, and check that the type of the subject (`IrGetValue`) is `UInt`? Are there cases where you would have unsafe-coerce from `UInt` to `Int` on LHS, but _not_ an unsafe-coerce on the RHS?

Reply:
I think the `isCoerce` check on the RHS as more of a precondition for accessing the constant expression that I expect to be there. On the way, it also checks the types.

I can't think of an example off the top of my head that does not conform to your second point, though. I think the front-end would need to put those coercions for it to type-check. Perhaps @sfs can think of a counter?

Reply:
These `unsafe-coerce` calls are put in by `JvmInlineClassLowering` in [specializeEqualsCall](https://github.com/JetBrains/kotlin/blob/3c8ef5749f3075c29f3a1a30b8a3b9c59cd7c73d/compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt#L330), so it should be fine. I'd still keep the check here, just in case.

Reply:
@sfs Thanks!